src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs

```diff
@@ -244,14 +262,102 @@ public void OnEventCommand(EventCommandEventArgs command)
                         return;
                     }
 #endif
-                    if (command.Command == EventCommand.Update || command.Command == EventCommand.Disable ||
-                        command.Command == EventCommand.Enable)
+
+                    string commandSessionId = GetSessionId(command);
+
+                    if ((command.Command == EventCommand.Update
```

Comment:
A lot of the actual code here needs to be cleaned up so don't bother too much with that, but here's the main idea:

- If we don't currently have a shared session, proceed as we would before
- If we have an active shared session, see if we can fuse with it (by checking that the args are consistent) -> if not, we give a `MultipleSessionsConfiguredIncorrectlyError`. -> if so, we call `Update` on the `_aggregationManager`.
- We do ref counting for enables/disables - once all SHARED sessions have been disabled, this frees the `_aggregationManager`, allowing a new session to start.

Reply:
Due to EventSource limitations I think you will struggle to do refcounting. I believe the only thing you can know for sure is that when IsEnabled() returns false then there are no sessions active. I think @davmason remembers these details better than I though.

(I ignored looking at the code in more detail, as you suggested)